File: banking_app_api/Services/CardService.cs
using Banking_App_API.Models.DTO;
using Banking_App_API.Repository;

namespace Banking_App_API.Services;

public class CardService(CardRepo repo)
{
    private readonly CardRepo _cardRepo = repo;

    public List<CardDto>? GetCardsByEmail(string? email) =>
        ValidationService.AllValid(email) && _cardRepo.GetCardsByEmail(email) is List<CardDto> cards ?
        cards : null;

    public CardDto? GetCardDetailsByEmailAndCardNo(string? email, string? cardNo) =>
        ValidationService.AllValid(email, cardNo) &&
        _cardRepo.GetCardDetailsByEmailAndCardNo(email, cardNo) is CardDto card ? card : null;

    public bool ResetCardPin(string? email, string? cardNo, string? oldPin, string? newPin) =>
        ValidationService.AllValid(email, cardNo, oldPin, newPin) &&
        _cardRepo.ResetCardPin(email, cardNo, oldPin, newPin);
}

File: banking_mobile_app/lib/services/card_service.dart
import 'dart:convert';

import 'package:banking_mobile_app/models/dto/card_dto.dart';
import 'package:banking_mobile_app/services/server_service.dart';
import 'package:http/http.dart' as http;

/// The app uses it to send HTTP requests to the API, concerning card related
/// matters.
class CardService {
  /// Fetches all the cards that the user has.
  static Future<List<CardDto>> getCardsByEmail(String email) async {
    http.Response resp = await http.get(
      Uri.parse(
        "${ServerService.serverIPAddressSN}/Card/ViewCardsByEmail/$email",
      ),
    );

    if (resp.statusCode == 200) {
      List<dynamic> jsonList = jsonDecode(resp.body);
      List<CardDto> cards = jsonList
          .map((card) => CardDto.fromJson(card))
          .toList();

      return cards;
    }
    throw http.ClientException("Failed to fetch cards");
  }

  /// Fetches the data for the selected card.
  static Future<CardDto> getCardDetailsByEmailAndCardNo(
    String email,
    String cardNo,
  ) async {
    http.Response resp = await http.get(
      Uri.parse(
        "${ServerService.serverIPAddressSN}/Card/ViewCardDetailsByEmailAndCardNo/$email/$cardNo",
      ),
    );

    if (resp.statusCode == 200) {
      Map<String, dynamic> json = jsonDecode(resp.body);
      return CardDto.fromJson(json);
    }
    throw http.ClientException("Failed to get card data.");
  }

  /// Resets the user's card pin.
  static Future<bool> resetCardPin(
    String email,
    String cardNo,
    String oldPin,
    String newPin,
  ) async {
    http.Response resp = await http.put(
      Uri.parse("${ServerService.serverIPAddressSN}/Card/ResetCardPin"),
      encoding: Encoding.getByName("utf-8"),
      headers: {"Content-Type": "application/json"},
      body: jsonEncode({
        "email": email,
        "cardNo": cardNo,
        "oldPin": oldPin,
        "newPin": newPin,
      }),
    );

    if (resp.statusCode == 200) {
      return true;
    }
    throw http.ClientException("Failed to reset card pin");
  }
}
